[OpenDataServices/cove#895] Better oneOf validation messages for records

Bjwebb opened the pull request:
OpenDataServices/cove#895

Demo of this http://895-oneof-messages.dev.cove.opendataservices.coop/review/?source_url=https://gist.githubusercontent.com/Bjwebb/f18545c007c31c6655eafbdcc1ff617c/raw/aa5a3e06b9baec1b8c5a2e1ffa4c86bdf3a63e66/records_invalid_releases_less_fuzzing_nested.json

odscjames commented on setup.cfg:
```diff
@@ -1,2 +1,4 @@
 [flake8]
 max-line-length=119
+# extend-ignore uses flake8's default ignore list, and adds to it
+extend-ignore = E501
```
Can we doc somewhere why we are adding this exception, and maybe discuss if we should? I tried to start this at https://opendataservices.gitbook.io/developer-docs/code/code-python#coding-style previously - otherwise it's really hard for new dev's to work out why we do certain things

Bjwebb replied:
Fair point. In this case it's related to a new formatting tool I was trying out ("black"). On reflection, I've decided to not make this change to the flake8 config now, and we can have a separate discussion about the new tool.

odscjames replied:
Thanks. I've heard of Black before and would be very interested to hear more, but that should def be a separate team discussion.

Bjwebb commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Data Review Tool: Correct reporting of large blocks of JSON](https://trello.com/c/dXzzBjKM/48-data-review-tool-correct-reporting-of-large-blocks-of-json)

odscjames commented:
So this sets the "assumption" key in the data? Does anything else happen in this library or does it happen elsewhere?

(and comment above - sorry, I know that's a small thing but it was exactly things like this that made it hard for me to know what was going on when I started so now we are more developers I'm keen to make sure we work well together!)

Bjwebb commented:
Nothing else happens in this library, but it will be used in the CoVE web UI https://github.com/OpenDataServices/cove/pull/1235/files

odscjames commented:
The changelog puzzles me - it says "Return specific validation errors for `oneOf` in OCDS records," but I can't see the bit where that happens. I can see it setting an assumption key in the return data, and in the other pull request I can see messages being shown to the user based on that. But I was expecting to then see something that loaded the specific bit of subschema and applied that to the JSON block, and that's how we get the more detailed error messages. Is that work done somewhere? But the code that sets assumptions and the tests for that look good, and if I've misunderstood the other bit I'm happy to approve them (maybe with a clearer changelog tho?)

      yield ValidationError(
                    "This array should contain either entirely embedded releases or "
                    "linked releases. Embedded releases contain an 'id' whereas linked "
                    "releases do not. Your releases contain a mixture."
                )

It would be good if we could attach some kind of data representation of this error to this so that anyone using the library to analyse results in bulk can easily do so without having to parse human strings (this use case has come up!) and to help with different UI's and I18N, but I understand ValidationError comes from another library so that's not easy.

Bjwebb commented:
The errors for the subschema are fetched at https://github.com/OpenDataServices/lib-cove/blob/a29683688288089dd9c7ceae539c2b212836307f/libcove/lib/common.py#L101 and then in my new code we iterate over `errs` and yield each error.

I've added an `error_id` to the `ValidationError`.

And, I've added a line about jsonschema version to the CHANGELOG.

odscjames commented:
The code looks good, thanks. 

Sorry for one last pedantic arse, should there be a new change log entry for " Add an `error_id` key to the validation JSON, ..."?

Bjwebb commented:
Sorry, my bad, that's an important one to capture. I've added it.
